### PR TITLE
plan: fix a bug when prune extra handle column

### LINKED ISSUE #4067: a DELETE statement using 'EXISTS' or 'IN' keyword, wrong records deleted
What did you do?
---
If possible, provide a recipe for reproducing the error.

```
SELECT id FROM t1;

id
---
123

SELECT id FROM t2;

id
---
123

SELECT id FROM t1
WHERE id IN (SELECT id FROM t2);

id
---
123

-- the statement should delete the record 123, but does not
DELETE FROM t1
WHERE id IN (SELECT id FROM t2);

-- should not delete the record, but the record gone
DELETE FROM t1
WHERE id NOT IN (SELECT id FROM t2);
```

I still can not reproduce the error, the error only happens in some of my tables.

What version of TiDB are you using (`tidb-server -V`)?
---

```
> $ ./tidb-server -V                                                                                                    
Git Commit Hash: bae8fe46f9305cdfff52cf18d5701ae7ead14d75
UTC Build Time:  2017-08-06 01:02:35


> $ ./tikv-server -V
TiKV 0.0.1
Git Commit Hash: c2d36f5b081339b24064c9630e876fcd8429746f
UTC Build Time:  2017-08-06 01:52:37
Rust Version:    1.19.0-nightly (d47cf08d5 2017-05-28)

> $ ./pd-server -V
Git Commit Hash: 30115ae539e7e175e988b8c70b630f0908d7a77b
UTC Build Time:  2017-08-04 12:16:13
```

The error occurs after we update to RC4.
Only DELETE statement, SELECT and UPDATE are fine.
The problem cause important data lost.

### PR DESCRIPTION
**Pull request by winoros**
If handle column is pruned, the map that stores the handle infomatioin should be cleared too.
This is to fix #4067 
PTAL @hanfei1991 @lamxTyler

### CODE REVIEW

**Comment by hanfei1991** [comment]
add some tests

File: plan/column_pruning.go [hunk 1]
```diff
@@ -157,13 +157,17 @@ func (p *Union) PruneColumns(parentUsedCols []*expression.Column) {
 // PruneColumns implements LogicalPlan interface.
 func (p *DataSource) PruneColumns(parentUsedCols []*expression.Column) {
 	used := getUsedList(parentUsedCols, p.schema)
+	handleIdx := -1 // -1 for not found.
```
Comment (jackysp) [inline plan/column_pruning.go]: It is better to define a const for handle index not found.

Comment (winoros) [inline plan/column_pruning.go]: This is only used in this method, i think we don't need to define one?

**Comment by hanfei1991** [comment]
LGTM

File: executor/write_test.go [hunk 2]
```diff
@@ -1087,3 +1087,22 @@ func assertEqualStrings(c *C, got []string, expect []string) {
 		c.Assert(got[i], Equals, expect[i])
 	}
 }
+
+// Test issue https://github.com/pingcap/tidb/issues/4067
```
Comment (lamxTyler) [inline executor/write_test.go]: Is this comment not needed? Because we can directly infer it from the function name.

Comment (winoros) [inline executor/write_test.go]: Many similar test method have similar comment. So i followed.

**Pull request merged**